bn-apps/cordapp-updates-distribution/README.md

```diff
@@ -0,0 +1,270 @@
+3. Add CorDapps that you would like to watch to `settings.yaml`
+4. Start using the utility by invoking `SYNC` or `PRINT_VERSIONS` commands
+
+# corda-updates-app
+
+`corda-updates-app`is a CorDapp that provides the following functionality:
+* A scheduled state to periodically check for new updates
+* Flows to schedule updates and retrieve available CorDapp versions
+* Flows to report CorDapp version information to the BNO
+
+## Scheduling synchronisation
+
+Updates can be scheduled via `ScheduleSyncFlow`. Synchronisation interval is driven by `syncInterval` configuration property and defaults to *5 hours*. `ScheduleSyncFlow` can be started from shell or RPC. 
+
+```kotlin
+// synchronisation should be launched asynchronously if corda-flows transport is used 
+subFlow(ScheduleSyncFlow(launchAsync = true))
+```
+
+> It is better to avoid launching ScheduleSyncFlow in synchronous mode, as it might result to a deadlock, as described in one of the previous sections
+
+## Getting available versions
+
+Available versions can be fetched via `GetAvailableVersionsFlow`. 
+
+```kotlin
+subFLow(GetAvailableVersionsFlow("net.corda:corda-finance"))
+```
+
+> GetAvailableVersionsFlow doesn't make any requests to the remote repository, but reuses a cache populated by ScheduleSyncFlow instead. At least one synchronisation should pass in order for GetAvailableVersionsFlow to return any results. 
+
+## Participant CorDapp Configuration
+
+Configuration is loaded from `cordapps/config/corda-updates-app.conf`.   
+
+```
+# path to the yaml configuration file (described in the previous sections) 
+configPath=./config.yaml
+
+# synchronisation interval in milliseconds defaults to 5 hours if not specified
+syncInterval=18000000
+
+# identity of the notary to use
+notary="O=Notary,L=London,C=GB"
+
+# identity of the BNO to report CorDapp versions to 
+bno="O=BNO,L=London,C=GB"
+```
+
+## Repository Hoster CorDapp Configuration
+
+Configuration is loaded from `cordapps/config/corda-updates-app.conf`. Repository hoster node is required only if `corda-flows` transport is used.
+```
+# URL of the repository where the hoster would serve the artifacts from. Supports http(s) and file transports. Proxies and authentication are not supported 
+remoteRepoUrl=file:/path/to/local/repository
+
+# Class that implements the SessionFilter interface. Optional.
+sessionFilter=com.my.app.MySessionFilter
+```
+## Stopping CorDapp from working if a newer version is available
+
+Such rules can be enforced in the following way:
+
+```kotlin
+class MyFlow : FlowLogic<Unit>() {
+    companion object {
+    // current version of CordDapp. Can be hardcoded as a constant or get received from serviceHub starting from Corda v4
+        val CURRENT_VERSION = "1.0"
+    }
+
+    @Suspendable
+    override fun call() {
+        val artifactMetadata = subFlow(GetAvailableVersionsFlow("my.cordapp.group:my-cordapp-id"))
```

Comment:
Ok, will remove this block. That particular requirement was flagged a couple of times in the groups.io thread.

Reply:
Yes the requirement to be able to approve, BUT, the problem is that the app has to keep working even if there's an unapproved update available. We shouldn't be killing flows with "please update", otherwise the moment the update is published there'd be an outage, right?

Reply:
Right, removed